src/adapter/src/config/frontend.rs

```diff
@@ -102,8 +111,25 @@ impl SystemParameterFrontend {
     pub async fn ensure_initialized(&self) {
         tracing::info!("waiting for SystemParameterFrontend to initialize");
 
-        // Start and initialize LD client for the frontend.
-        self.ld_client.start_with_default_executor();
+        // Start and initialize LD client for the frontend. The callback passed
+        // will export the last time when an SSE event from the LD server was
+        // received in a Prometheus metric.
+        self.ld_client.start_with_default_executor_and_callback({
+            let last_sse_time_seconds = self.ld_metrics.last_sse_time_seconds.clone();
```

Comment:
I don't understand what we need this block for. Is the very first `t_curr` here 0? Which means we'll inc_by the current unix epoch, but then in the loop above increment by something in the tens or hundreds?

What happens if we don't have this block at all?

Reply:
If you mean the `if let` block - that is needed because `duration_since` will only a `Duration` if the difference is positive.

The inner `if` block is needed because we cannot call `inc_by` with a negative number.

I can probably also initialize `last_sse_time_seconds` to the current unix timestamp and the ln use it to obtain a base timestamp for the `duration_since` call whenever the callback function is invoked, but this will come at the price of a stack allocation for the `SystemTime` derived from the current `last_sse_time_seconds` value.

Reply:
I'm confused about the entire closure in start_with_default_executor_and_callback. Is it called one time or multiple? I don't understand why we need this closure and the code above that also sets this metric.

What is `t_curr`'s value for the first call to `last_sse_time_seconds.get()`?

Calling `SystemTime::now` is kinda unfortunate. We've successfully avoided doing that except in exactly one place by plumbing down a `NowFn` everywhere, which makes it possible to test things that care about time and also more easily audit when time is fetched.

Reply:
>  I don't understand why we need this closure and the code above that also sets this metric.

There are two `last_known_time` metrics that we want to export. 

1. The first is for the events that the client sends to the server (this is the old callback, with now provisioned new metric name `last_cse_time_seconds`).
2. The second is for events received by the client from the server (this is the callback that you are looking at with updates the `last_sse_time_seconds` metric).

Once we have the PR merged I will update our health dashboard with a panel for (2) in addition to the panel for (1).

> Is it called one time or multiple?

This is going to be called every time when a server-side event is received. I had to tweak `rust-server-sdk` to add support for these callbacks, and in order to minimize the amount of changes I opted to not put both in the same place. You can review the last two commits [of our `rust-server-sdk` fork](https://github.com/MaterializeInc/rust-server-sdk/commits/main) for details.

> What is `t_curr`'s value for the first call to `last_sse_time_seconds.get()`?

The default value for the `IntCounter` is `0`.

> Calling `SystemTime::now` is kinda unfortunate.

I'll see if I can make it work with `NowFn`.